tokio/src/task/local.rs

```diff
@@ -693,21 +706,27 @@ impl LocalSet {
     fn with_if_possible<T>(&self, f: impl FnOnce() -> T) -> T {
         let mut f = Some(f);
 
-        let res = CURRENT.try_with(|LocalData { ctx, .. }| {
+        let res = CURRENT.try_with(|LocalData { ctx, entered, .. }| {
             struct Reset<'a> {
                 ctx_ref: &'a RcCell<Context>,
-                val: Option<Rc<Context>>,
+                entered_ref: &'a Cell<bool>,
+                ctx_val: Option<Rc<Context>>,
+                entered_val: bool,
             }
             impl<'a> Drop for Reset<'a> {
                 fn drop(&mut self) {
-                    self.ctx_ref.replace(self.val.take());
+                    self.ctx_ref.set(self.ctx_val.take());
+                    self.entered_ref.set(self.entered_val)
                 }
             }
-            let old = ctx.replace(Some(self.context.clone()));
+            let ctx_old = ctx.replace(Some(self.context.clone()));
+            let entered_old = entered.replace(false);
 
             let _reset = Reset {
                 ctx_ref: ctx,
-                val: old,
+                entered_ref: entered,
+                ctx_val: ctx_old,
+                entered_val: entered_old,
```

Comment:
take it or leave it: the `Reset` code got substantially more complex in this change, and it's used in two different places. now that it's no longer fairly trivial, what do you think about moving this out of the `with` and `with_if_possible` functions, so that we can use one implementation in both places. Something like:
```rust
impl LocalData {
    fn enter(&self, context: Rc<Context>) -> Reset<'_> {
        let ctx_val = self.ctx.replace(Some(context));
        let entered_val = self.entered.replace(false);

        Reset {
            ctx_ref: self.ctx,
            entered_ref: self.entered,
            ctx_val: ctx_old,
            entered_val: entered_old,
        }
    }
}

struct Reset<'a> {
    ctx_ref: &'a RcCell<Context>,
    entered_ref: &'a Cell<bool>,
    ctx_val: Option<Rc<Context>>,
    entered_val: bool,
}

impl<'a> Drop for Reset<'a> {
    fn drop(&mut self) {
        self.ctx_ref.set(self.ctx_val.take());
        self.entered_ref.set(self.entered_val)
    }
}
```

and then changing `with` and `with_if_possible` to just call
```rust
let _reset = local_data.enter(self.context.clone());
```

Reply:
Thank you! I've moved out, and renamed it to more verbose name.